docs/spirv-fuzz.md

```diff
@@ -0,0 +1,83 @@
+# Guide to writing a spirv-fuzz fuzzer pass
+
+Writing a spirv-fuzz fuzzer pass usually requires two main contributions:
+
+- A *transformation*, capturing a small semantics-preserving change that can be made to a SPIR-V module.  This requires adding a protobuf message representing the transformation, and a corresponding class that implements the `Transformation` interface.
+- A new *fuzzer pass* class, implementing the `FuzzerPass` interface, that knows how to walk a SPIR-V module and apply the new transformation in a randomized fashion.
+
+In some cases, more than one kind of transformation is required for a single fuzzer pass, and in some cases the transformations that a new fuzzer pass requires have already been introduced by existing passes.  But the most common case is to introduce a transformation and fuzzer pass together.
+
```

Comment:
Note that all fuzzer passes _record_ the transformations as they are applied, resulting in a list of transformation protobuf messages. If we take the original SPIR-V module (without any transformations applied), we can re-apply all of the transformations in the list (in the same order) to get to the same transformed SPIR-V module.

Crucially, we can _replay_ a _subset_ of the transformations. Thus, when replaying an individual transformation, the SPIR-V module might be _different_ from when the transformation was created because earlier transformations were not replayed. As we will see, transformations much be designed with this in mind.